pkg/registry/rbac/reconciliation/reconcile_clusterrole.go

```diff
@@ -0,0 +1,188 @@
+	ReconcileUpdate   ReconcileOperation = "update"
+	ReconcileRecreate ReconcileOperation = "recreate"
+	ReconcileNone     ReconcileOperation = "none"
+)
+
+type ReconcileClusterRoleOptions struct {
+	// Role is the expected role that will be reconciled
+	Role *rbac.ClusterRole
+	// Confirm indicates writes should be performed. When false, results are returned as a dry-run.
+	Confirm bool
+	// Union indicates reconciliation should be additive (no permissions should be removed)
+	Union bool
+	// Client is used to look up existing roles, and create/update the role when Confirm=true
+	Client internalversion.ClusterRoleInterface
+}
+
+type ReconcileClusterRoleResult struct {
+	// Role is the resulting role from the reconciliation operation
+	Role *rbac.ClusterRole
+
+	// MissingRules contains expected rules that were missing from the currently persisted role
+	MissingRules []rbac.PolicyRule
+	// ExtraRules contains extra permissions the currently persisted role had
+	ExtraRules []rbac.PolicyRule
+
+	// Operation is the API operation required to reconcile
+	Operation ReconcileOperation
+	// Protected indicates an existing role prevented reconciliation
+	Protected bool
+}
+
+func (o *ReconcileClusterRoleOptions) Run() (*ReconcileClusterRoleResult, error) {
+	var result *ReconcileClusterRoleResult
+
+	existing, err := o.Client.Get(o.Role.Name, metav1.GetOptions{})
+	switch {
+	case errors.IsNotFound(err):
+		result = &ReconcileClusterRoleResult{
+			Role:         o.Role,
+			MissingRules: o.Role.Rules,
+			Operation:    ReconcileCreate,
+		}
+
+	case err != nil:
+		return nil, err
+
+	default:
+		result, err = changedRole(existing, o.Role, o.Union)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// If reconcile-protected, short-circuit
+	if result.Protected {
+		return result, nil
+	}
+	// If we're in dry-run mode, short-circuit
+	if !o.Confirm {
+		return result, nil
+	}
+
+	switch result.Operation {
+	case ReconcileCreate:
+		created, err := o.Client.Create(result.Role)
+		// If created since we started this reconcile, re-run
+		if errors.IsAlreadyExists(err) {
+			return o.Run()
+		}
+		if err != nil {
+			return nil, err
+		}
+		result.Role = created
+
+	case ReconcileUpdate:
+		updated, err := o.Client.Update(result.Role)
+		// If deleted since we started this reconcile, re-run
+		if errors.IsNotFound(err) {
+			return o.Run()
+		}
+		if err != nil {
+			return nil, err
+		}
+		result.Role = updated
+
+	case ReconcileNone:
+		// no-op
+
+	default:
+		return nil, fmt.Errorf("invalid operation: %v", result.Operation)
+	}
+
+	return result, nil
+}
+
+// ChangedClusterRoles returns the roles that must be created and/or updated to
+// match the recommended bootstrap policy
+func changedRole(existing, expected *rbac.ClusterRole, union bool) (*ReconcileClusterRoleResult, error) {
+	result := &ReconcileClusterRoleResult{Operation: ReconcileNone}
+
+	result.Protected = (existing.Annotations[ReconcileProtectAnnotation] == "true")
```

Comment:
Still handy to see what the diff is, since you have to manage adding expected permissions yourself